specs/ChainSpec/Upgradability.md

```diff
@@ -0,0 +1,71 @@
+# Upgradability
+
+This part of specification describes specifics of upgrading the protocol, and touches on few different parts of the system.
+
+High level upgradability consists of three parts:
+1. Updating without any changes to underlaying data structures or protocol;
+2. Updating when underlaying data structures changed (config, database or something else internal to the node and probably client specific);
+3. Updating with protocol changes that all validating nodes must adjust to.
+
+## Versioning
+
+There are 2 different important versions:
+- Version of binary defines it's internal data structures / database and configs. This version is client specific and doesn't need to be matching between nodes.
+- Version of the protocol, defining the "language" nodes are speaking.
+
+```rust
+/// Latest version of protocol that this binary can work with.
+type ProtocolVersion = u32;
+```
+
+Open questions: do we want to use something better than a number of versioning protocol?
+Ideally it should be some hash of protocol spec that this binary supports, but we don't have a canonical protocol spec versioning yet. We can use commit hash of `spec` / `NEPs` repo.
+
+## Client versioning
+
+Clients should follow [semantic versioning](https://semver.org/).
+Specifically:
+ - MAJOR version defines protocol releases.
+ - MINOR version defines changes that are client specific but require database migration, change of config or something similar. This includes client-specific features. Client should execute migrations on start, by detecting that information on disk is produced by previous version and auto-migrate it to new one.
+  - PATCH version defines when bug fixes, which should not require migrations or protocol changes.
+
+Clients can define how current version of data is stored and migrations applied.
+General recommendation is to store version in the database and on binary start, check version of database and perform required migrations.
+
+## Protocol Upgrade
+
+### Block structure
+
+```rust
+struct BlockHeaderInnerRest {
+    ...
+    version: ProtocolVersion,
+}
+```
+
+Each validator when producing block includes latest version their node supports.
+
+### Consensus
+
+| Name | Value |
+| - | - |
+| `PROTOCOL_UPGRADE_BLOCK_THRESHOLD` | `80%` |
+
+The condition to switch to next protocol version is based on % of blocks in previous epoch:
```

Comment:
As discussed in the meeting the other day, we may want to change this from blocks being labelled with a version number to having the upcoming block producers declare a version they wish to support in the next epoch and set the epoch's version based on this vote (we could probably keep the 80% threshold).

I think even in this model it is a good idea to include the version number in the block header because this will allow handling messages from previous epochs properly.

One question I have about any model of upgrades is what happens to the block producers that are not yet ready? This could be up to 20% (by stake) of the block producers. Will they simply not produce blocks until they upgrade? If they do make blocks according to the last version do we still accept them or will they be considered invalid?

Reply:
The problem with upcoming block producers declaring is:
 - They would need to explicitly send staking transaction (which might require some HSM / offline / multisig custody signing which is hard)
 - Validators will be unstaking a bit of stake, because when they send re-staking transaction they need to specify the amount with the future reward to match what we do by default in rollovers. 
 - Also problem is staking pool, as we need to change version inside them to allow them to send staking transactions with new version.

So alternative is to create new Transaction Action to `VersionVote(new-version)`, that can be signed with staking key (which is not supported by any of our current infra right now) and that node automatically sends and this gets attached to proposals -> epoch manager.

Reply:
> So alternative is to create new Transaction Action to VersionVote(new-version), that can be signed with staking key (which is not supported by any of our current infra right now) and that node automatically sends and this gets attached to proposals -> epoch manager.

Also we can just add `protocol_version` into the StakeAction
https://github.com/nearprotocol/nearcore/blob/master/core/primitives/src/transaction.rs#L124-L132

Reply:
@fckt that's what I argued against above the alternative option.